kratos/processes/apply_periodic_boundary_condition_process.cpp

```diff
@@ -149,92 +150,125 @@ template <int TDim>
-    const auto& r_var_z = KratosComponents<VariableType>::Get(rVarName + std::string("_Z"));
+    const auto& r_var_x = (*rVars[0]);
+    const auto& r_var_y = (*rVars[1]);
+    const auto& r_var_z = (*rVars[2]);
 
     // Reference constraint
-    const auto& r_clone_constraint = KratosComponents<MasterSlaveConstraint>::Get("LinearMasterSlaveConstraint");
+    const auto& r_clone_constraint = LinearMasterSlaveConstraint();
 
     IndexType master_index = 0;
-    for (auto& master_node : rHostedGeometry)
+    for (auto& r_master_node : rHostedGeometry)
     {
         const double master_weight = rWeights(master_index);
 
         const double constant_x = master_weight * mTransformationMatrixVariable(0,3);
         const double constant_y = master_weight * mTransformationMatrixVariable(1,3);
-        const double constant_z = master_weight * mTransformationMatrixVariable(2,3);
-
-        #pragma omp critical
-        {
-            int current_num_constraint = mrMasterModelPart.GetRootModelPart().NumberOfMasterSlaveConstraints();
-            auto constraint1 = r_clone_constraint.Create(++current_num_constraint, master_node, r_var_x, rSlaveNode, r_var_x, master_weight * mTransformationMatrixVariable(0,0), constant_x);
-            auto constraint2 = r_clone_constraint.Create(++current_num_constraint, master_node, r_var_y, rSlaveNode, r_var_x, master_weight * mTransformationMatrixVariable(0,1), constant_x);
-            auto constraint3 = r_clone_constraint.Create(++current_num_constraint, master_node, r_var_z, rSlaveNode, r_var_x, master_weight * mTransformationMatrixVariable(0,2), constant_x);
 
-            auto constraint4 = r_clone_constraint.Create(++current_num_constraint, master_node, r_var_x, rSlaveNode, r_var_y, master_weight * mTransformationMatrixVariable(1,0), constant_y);
-            auto constraint5 = r_clone_constraint.Create(++current_num_constraint, master_node, r_var_y, rSlaveNode, r_var_y, master_weight * mTransformationMatrixVariable(1,1), constant_y);
-            auto constraint6 = r_clone_constraint.Create(++current_num_constraint, master_node, r_var_z, rSlaveNode, r_var_y, master_weight * mTransformationMatrixVariable(1,2), constant_y);
+        if constexpr (TDim == 3) {
+            const double constant_z = master_weight * mTransformationMatrixVariable(2,3);
 
-            mrMasterModelPart.AddMasterSlaveConstraint(constraint1);
-            mrMasterModelPart.AddMasterSlaveConstraint(constraint2);
-            mrMasterModelPart.AddMasterSlaveConstraint(constraint3);
-            mrMasterModelPart.AddMasterSlaveConstraint(constraint4);
-            mrMasterModelPart.AddMasterSlaveConstraint(constraint5);
-            mrMasterModelPart.AddMasterSlaveConstraint(constraint6);
-
-
-            if (TDim == 3) // TODO: This function can be optimized using template specialization. if constexpr will fail due to unused variables declared
+            #pragma omp critical
```

Comment:
@philbucher do we have something in mind to emulate ```#pragma omp critical``` generically in ```ParallelUtilities```?

Reply:
this bad boy:
https://github.com/KratosMultiphysics/Kratos/blob/fef4fb6be42bf22839de8219c1fc61eef9e8d795/kratos/sources/parallel_environment.cpp#L249

Although generally I recommend to use reductions (which internally would use the same)

Reply:
but ```LockObject``` is not implemented for *C++11* parallelism (yet?) right?

Reply:
almost nothing is fully functional with C++11 based parallelism
The big fat warning in CMake is there for a reason :D

But hopefully soon it will